hashintel/hash
# H-2453, H-2479: initial backend implementation of HASH Flows

## 🌟 What is the purpose of this PR?




This PR introduces adds an initial implementation of `HASH Flows`. Currently flows run in a temporal `runFlow` workflow, which calls the relevant actions in the flow DAG as temporal activities concurrently when their dependencies have been met.

The initial set of "actions" introduced of this PR (steps that can be used in a flow) include:
- `generateWebQueries`
- `webSearch`
- `getWebPageByUrl`
- `inferEntitiesFromContent`
- `persistEntity`
- `getFileFromUrl`
- `researchEntities`
- `getWebPageSummary`

## 🔗 Related links





- H-2453
- H-2479
- [Document specifying the initial implementation](https://www.notion.so/hashintel/Flows-Implementation-1e0a55c0a0d94b4c80866739bd2d45c8) (_internal link_)

## 🔍 What does this change?




- adds TS types for `FlowDefinition`s, and `Flow` runs
- adds system types for `Flow Definition`s and `Flow` runs (note these will not be seeded in production for the time being. To use them run HASH in the developer environment)
- a `startFlow` GQL resolver, which accepts a flow definition and a trigger and executes the flow in temporal
- a `runFlow` temporal workflow, which runs the actions in a flow, persists a `Flow` entity in the graph representing the flow, and returns its outputs
- replaces the existing "research task" temporal workflow with a HASH flow composed of actions

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:


### 📜 Does this require a change to the docs?




The changes in this PR:


### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## ⚠️ Known issues



`HASH Flows` is not ready for production, some of the key blockers for include:
- tracking OpenAi usage incurred by a user when executing a flow run
- stabilising the types to avoid future migrations

## 🐾 Next steps



- Continue work on the infer entities from web page worker agent, so it can handle more types (H-2515)
- Unify logging approach with the one that will added in [this follow-up PR](https://github.com/hashintel/hash/pull/4290)
- Support link inference in flows (H-2496)

## 🛡 What tests cover this?



None so far.

## ❓ How to test this?



Either use the "AI" page:
1. Visit `http://localhost:3000/ai`, and try out the research task with a simple entity type and prompt
2. Observe the logs from the AI temporal worker to inspect progress
3. If the run is successful, you will see a list of inferred entities from the web based on your prompt

Or run a custom flow via the `startFlow` GQL resolver:
1. Define a flow definition according to the `FlowDefinition` TS type
2. Pass the flow definition and the trigger to the `startFlow` GQL resolver
3. Observe logs from the AI temporal worker to inspect progress
4. Observe the outputs in the resolver response
